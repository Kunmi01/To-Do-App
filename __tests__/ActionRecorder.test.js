import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import ConnectedActionRecorder, {
  ActionRecorder
} from '../src/components/ActionRecorder';
import {
  startRecordingActions,
  stopRecordingActions,
  playRecordedActions,
  clearRecordedActions
} from '../src/redux/actions';

describe('ActionRecorder', () => {
  const initialState = {
    isRecording: false,
    isPlaying: false,
    recordedActions: []
  };
  const mockStore = configureStore();
  let store;
  let container;
  let actionRecorderContainer;

  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
    container = mount(
      <Provider store={store}>
        <ConnectedActionRecorder />
      </Provider>
    );
    actionRecorderContainer = container.find(ActionRecorder);
  });

  it('renders connected(SMART) component', () => {
    expect(container.find(ConnectedActionRecorder).exists()).toBe(true);
    expect(toJson(container)).toMatchSnapshot();
  });

  it('props match with initialState', () => {
    expect(actionRecorderContainer.prop('isRecording')).toEqual(
      initialState.isRecording
    );
    expect(actionRecorderContainer.prop('isPlaying')).toEqual(
      initialState.isPlaying
    );
    expect(actionRecorderContainer.prop('recordedActions')).toEqual(
      initialState.recordedActions
    );
  });

  it('can RECORD actions', () => {
    renderer.act(() => {
      actionRecorderContainer
        .find('.action-recorder__button--record')
        .simulate('click');
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(startRecordingActions());
  });

  it('can STOP recording actions', () => {
    store = mockStore({
      isRecording: true,
      isPlaying: false,
      recordedActions: []
    });
    store.dispatch = jest.fn();
    container = mount(
      <Provider store={store}>
        <ConnectedActionRecorder />
      </Provider>
    );
    actionRecorderContainer = container.find(ActionRecorder);

    renderer.act(() => {
      actionRecorderContainer
        .find('.action-recorder__button--stop')
        .simulate('click');
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(stopRecordingActions());
  });

  it('can PLAY recorded actions', () => {
    store = mockStore({
      isRecording: false,
      isPlaying: false,
      recordedActions: [{ type: 'test type', payload: { itemId: 'test id' } }]
    });
    store.dispatch = jest.fn();
    container = mount(
      <Provider store={store}>
        <ConnectedActionRecorder />
      </Provider>
    );
    actionRecorderContainer = container.find(ActionRecorder);

    renderer.act(() => {
      actionRecorderContainer
        .find('.action-recorder__button--play')
        .simulate('click');
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(playRecordedActions());
  });

  it('can CLEAR recorded actions', () => {
    store = mockStore({
      isRecording: false,
      isPlaying: false,
      recordedActions: [{ type: 'test type', payload: { itemId: 'test id' } }]
    });
    store.dispatch = jest.fn();
    container = mount(
      <Provider store={store}>
        <ConnectedActionRecorder />
      </Provider>
    );
    actionRecorderContainer = container.find(ActionRecorder);

    renderer.act(() => {
      actionRecorderContainer
        .find('.action-recorder__button--clear')
        .simulate('click');
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(clearRecordedActions());
  });
});
