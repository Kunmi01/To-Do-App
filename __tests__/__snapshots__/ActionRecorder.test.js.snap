// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionRecorder renders connected(SMART) component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ActionRecorder)>
    <ActionRecorder
      dispatch={[MockFunction]}
      isPlaying={false}
      isRecording={false}
      recordedActions={Array []}
    >
      <div
        className="action-recorder"
      >
        <div
          className="action-recorder__wrapper"
        >
          <h3
            className="action-recorder__header"
          >
            Action Recorder
          </h3>
          <div
            className="action-recorder__inner-wrapper"
          >
            <div
              className="action-recorder__buttons-wrapper"
            >
              <button
                className="action-recorder__button action-recorder__button--record"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Record
              </button>
              <button
                className="action-recorder__button action-recorder__button--stop"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                Stop
              </button>
              <button
                className="action-recorder__button action-recorder__button--play"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                Play
              </button>
              <button
                className="action-recorder__button action-recorder__button--clear"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Clear
              </button>
            </div>
            <p
              className="action-recorder__status"
            >
              <span>
                status: 
              </span>
              Stopped
            </p>
            <div
              className="action-recorder__feedback-wrapper"
            />
          </div>
        </div>
      </div>
    </ActionRecorder>
  </Connect(ActionRecorder)>
</Provider>
`;
